lib/ansible/modules/clustering/kubevirt/kubevirt_cdi_upload.py

```diff
@@ -0,0 +1,205 @@
+    pvc_name: pvc-vm1
+    upload_host: https://localhost:8443
+    upload_host_verify_ssl: false
+    path: /tmp/cirros-0.4.0-x86_64-disk.img
+'''
+
+RETURN = '''# '''
+
+import copy
+import traceback
+import requests
+
+from collections import defaultdict
+
+from ansible.module_utils.k8s.common import AUTH_ARG_SPEC
+from ansible.module_utils.k8s.raw import KubernetesRawModule
+
+
+SERVICE_ARG_SPEC = {
+    'pvc_name': {'required': True},
+    'pvc_namespace': {'required': True},
+    'upload_host': {'required': True},
+    'upload_host_verify_ssl': {
+        'type': 'bool',
+        'default': True
+    },
+    'path': {'required': True},
+    # Workaround for ansible <2.8
+    'kind': {
+        'default': 'UploadTokenRequest',
+        'choices': ['UploadTokenRequest'],
+    },
+    'api_version': {
+        'default': 'upload.cdi.kubevirt.io/v1alpha1',
+        'choices': ['upload.cdi.kubevirt.io/v1alpha1'],
+    },
+    'name': {
+        'default': None,
+    },
+    'namespace': {
+        'default': None,
+    },
+    'src': {
+        'default': None,
+    },
+    'resource_definition': {
+        'default': None,
+    },
+    'merge_type': {
+        'type': 'list',
+    }
+}
+
+
+class KubeVirtCDIUpload(KubernetesRawModule):
+    def __init__(self, *args, **kwargs):
+        super(KubeVirtCDIUpload, self).__init__(*args, **kwargs)
+
+    @property
+    def argspec(self):
+        """ argspec property builder """
+        argument_spec = copy.deepcopy(AUTH_ARG_SPEC)
+        argument_spec.update(SERVICE_ARG_SPEC)
+        return argument_spec
+
+    def execute_module(self):
+        """ Module execution """
+        self.client = self.get_api_client()
+
+        api_version = 'upload.cdi.kubevirt.io/v1alpha1'
+        pvc_name = self.params.get('pvc_name')
+        pvc_namespace = self.params.get('pvc_namespace')
+        upload_host = self.params.get('upload_host')
+        upload_host_verify_ssl = self.params.get('upload_host_verify_ssl')
+        path = self.params.get('path')
+
+        definition = defaultdict(defaultdict)
+
+        definition['kind'] = 'UploadTokenRequest'
+        definition['apiVersion'] = api_version
+
+        def_meta = definition['metadata']
+        def_meta['name'] = pvc_name
+        def_meta['namespace'] = pvc_namespace
+
+        def_spec = definition['spec']
+        def_spec['pvcName'] = pvc_name
+
+        # Let's check the file's there before we do anything else
+        imgfile = open(path, 'rb')
+
+        resource = self.find_resource('UploadTokenRequest', api_version, fail=True)
+        definition = self.set_defaults(resource, definition)
+        result = self.perform_action(resource, definition)
+
+        headers = {'Authorization': "Bearer {}".format(result['result']['status']['token'])}
+        files = {'file': imgfile}
+        url = "{}/v1alpha1/upload".format(upload_host)
+        requests.post(url, files=files, headers=headers, verify=upload_host_verify_ssl)
```

Comment:
Should we have wait here true/false?

Reply:
Unless you mean something else, it's not possible to have a non–wait upload code. By definition it needs to be running while the file is being uploaded and then terminates once everything gets uploaded.